subsys/net/lib/coap/coap_client.c

```
@@ -0,0 +1,639 @@
+		return ret;
+	}
+
+	ret = sendto(fds.fd, request.data, request.offset, 0, &address, sizeof(address));
+	if (ret < 0) {
+		LOG_ERR("Error sending CoAP reset message");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int handle_response(const struct coap_packet *response)
+{
+	int ret = 0;
+	int response_type;
+	static int offset;
+	int block_option;
+	int block_num;
+	bool blockwise_transfer = false;
+	bool last_block = false;
+
+	/* Handle different types, ACK might be separate or piggybacked
+	 * CON and NCON contains a separate response, CON needs an empty response
+	 * CON request results as ACK and possibly separate CON or NCON response
+	 * NCON request results only as a separate CON or NCON message as there is no ACK
+	 * With RESET, just drop gloves and call the callback.
+	 */
+	response_type = coap_header_get_type(response);
+
+	/* Reset and Ack need to match the message ID with request */
+	if ((response_type == COAP_TYPE_ACK || response_type == COAP_TYPE_RESET) &&
+	     coap_header_get_id(response) != pending.id)  {
+		LOG_ERR("Unexpected ACK or Reset");
+		return -EFAULT;
+	} else if (response_type == COAP_TYPE_RESET) {
+		coap_pending_clear(&pending);
+	}
+
+	/* CON, NON_CON and piggybacked ACK need to match the token with original request */
+	uint16_t payload_len;
+	uint8_t response_code = coap_header_get_code(response);
+	const uint8_t *payload = coap_packet_get_payload(response, &payload_len);
+
+	/* Separate response */
+	if (payload_len == 0 && response_type == COAP_TYPE_ACK && response_code == 0) {
+		/* Set a timeout value and clear pending */
+		coap_pending_clear(&pending);
+		return 1;
+	}
+
+	/* Check for tokens */
+	if (!token_compare(&request, response)) {
+		LOG_ERR("Not matching tokens, respond with reset");
+		ret = send_reset(response, COAP_RESPONSE_CODE_NOT_FOUND);
+		return 1;
+	}
+
+	/* Send ack for CON */
+	if (response_type == COAP_TYPE_CON) {
+		ret = send_ack(response, COAP_RESPONSE_CODE_OK);
+		if (ret < 0) {
+			return ret;
```

maybe `goto fail` instead of returning here?